### PR TITLE
Added the example guideline and template

### PR DESCRIPTION
**Pull request by pedrominatel**
## Description of Change
This PR adds the example guideline and template.

## Tests scenarios
N/A

## Related links

https://github.com/espressif/arduino-esp32/issues/7659

### CODE REVIEW

**Review by mrengineer7777 — COMMENTED** [review]
LGTM

File: docs/source/contributing.rst [hunk 1]
```diff
@@ -39,11 +39,95 @@ Once the Pull Request is ready to merge, it will first be merged into our intern
...
+* Use the `Arduino style guide`_
+* Add the header to all source files
+* Add the `README.md` file
+* Add inline comments if needed
+* Test the example
+
+Header
+******
+
+All the source files must include the header with the example name and license, if applicable. You can change this header as you with, but it will be reviewed by the community and may not be accepted.
```
Comment (P-R-O-C-H-Y) [inline docs/source/contributing.rst:58]: ... You can change this header as you **wish**, ...

File: docs/source/contributing.rst [hunk 2]
```diff
@@ -39,11 +39,95 @@ Once the Pull Request is ready to merge, it will first be merged into our intern
...
+***********
+
+The **README.md** file should contain the example details.
+
+Please see the recommended **README.md** file in the `example template folder <https://github.com/espressif/arduino-esp32/tree/master/libraries/ESP32/examples/Template/ExampleTemplate>`_.
+
+Inline Comments
+***************
+
+Inline comments are important if the example contains complex algorithms os specific configuration that the user needs to change.
```
Comment (P-R-O-C-H-Y) [inline docs/source/contributing.rst:86]: ... complex algorithms **or** specific ...

File: docs/source/contributing.rst [hunk 3]
```diff
@@ -24,7 +24,7 @@ Before sending us a Pull Request, please consider this:
 * Are comments and documentation written in clear English, with no spelling or grammar errors?
 
 * Example contributions are also welcome.
-  * If you are contributing by adding a new example, please use the `Arduino style guide`_.
+  * If you are contributing by adding a new example, please use the `Arduino style guide`_ and the example guideline below.
```
Comment (P-R-O-C-H-Y) [inline docs/source/contributing.rst:27]: Missing empty line between:
`* Example contributions are also welcome.`
and
`* If you are contributing by adding a new example, please use the...`

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
PTAL @pedrominatel.

**Comment by pedrominatel** [comment]
Thanks for the review, @P-R-O-C-H-Y.

**Review by P-R-O-C-H-Y — APPROVED** [review]
LGTM!

File: docs/source/contributing.rst [hunk 4]
```diff
@@ -15,16 +15,17 @@ Before Contributing
 
 Before sending us a Pull Request, please consider this:
 
-* Is the contribution entirely your own work, or is it already licensed under an LGPL 2.1 compatible Open Source License? If not, we unfortunately cannot accept it.
+* Is the contribution entirely your own work, or is it already licensed under an LGPL 2.1 compatible Open Source License? If not, we, unfortunately, cannot accept it.
```
Comment (VojtechBartoska) [inline docs/source/contributing.rst:18]: This sentence is a bit awkward.

Suggestion:
If it's not, we unfortunately cannot accept it.

File: docs/source/contributing.rst [hunk 5]
```diff
@@ -15,16 +15,17 @@ Before Contributing
...
 
-* Is the code adequately commented for people to understand how it is structured?
+* Is the code adequately commented on to people understand how it is structured?
```
Comment (VojtechBartoska) [inline docs/source/contributing.rst:20]: Again, this sentence is a bit awkward.

Suggestion:

Is the code adequately commented so others can understand how it is structured?

File: docs/source/contributing.rst [hunk 6]
```diff
@@ -33,17 +34,101 @@ If you're unsure about any of these points, please open the Pull Request anyhow
 Pull Request Process
 --------------------
 
-After you open the Pull Request, there will probably be some discussion in the comments' field of the request itself.
+After you open the Pull Request, there will probably be some discussion in the comments field of the request itself.
 
 Once the Pull Request is ready to merge, it will first be merged into our internal git system for in-house automated testing.
```
Comment (VojtechBartoska) [inline docs/source/contributing.rst:39]: Optionally, we can put phase "in-house" into quotation marks as it's our "naming".

File: docs/source/contributing.rst [hunk 7]
```diff
@@ -33,17 +34,101 @@ If you're unsure about any of these points, please open the Pull Request anyhow
...
 
-Legal Part
-----------
+Example Contribution Guideline
+------------------------------
+
+Checklist
+*********
+
+* Check if your example proposal has no similarities to the project
```
Comment (VojtechBartoska) [inline docs/source/contributing.rst:49]: I will add here this to make it more clear:

* Check if your example proposal has no similarities to the project **(already existing examples)**

File: docs/source/contributing.rst [hunk 8]
```diff
@@ -33,17 +34,101 @@ If you're unsure about any of these points, please open the Pull Request anyhow
...
+* Add the `README.md` file
+* Add inline comments if needed
+* Test the example
+
+Header
+******
+
+All the source files must include the header with the example name and license, if applicable. You can change this header as you wish, but it will be reviewed by the community and may not be accepted.
+
+Additionally, you can add some description about the example, links to the documentation, or the author's name. Just have in mind to keep it simple and short.
```
Comment (VojtechBartoska) [inline docs/source/contributing.rst:61]: I will change here the word of Additionally to "Ideally" :) 

Maybe add link to "readme.md" of the example?

File: docs/source/contributing.rst [hunk 9]
```diff
@@ -33,17 +34,101 @@ If you're unsure about any of these points, please open the Pull Request anyhow
...
+***********
+
+The **README.md** file should contain the example details.
+
+Please see the recommended **README.md** file in the `example template folder <https://github.com/espressif/arduino-esp32/tree/master/libraries/ESP32/examples/Template/ExampleTemplate>`_.
+
+Inline Comments
+***************
+
+Inline comments are important if the example contains complex algorithms or specific configurations that the user needs to change.
```
Comment (VojtechBartoska) [inline docs/source/contributing.rst:88]: What about addding here a link to nicely commented example?

Comment (VojtechBartoska) [inline docs/source/contributing.rst:88]: _I will add one more sentence:_

Brief and clear inline comments are really helpful for the example understanding and it's fast usage.

**Review by VojtechBartoska — APPROVED** [review]
In general, the proposal is nice. I'm sharing just few more likely format and language suggestions.

**Comment by pedrominatel** [comment]
PTAL, @VojtechBartoska and merge if everything is ok.

File: libraries/ESP32/examples/Template/ExampleTemplate/README.md [hunk 10]
```diff
@@ -98,21 +98,21 @@ Chip ID: 3957392
...
-Before creating a new issue, be sure to try the Troubleshooting and to check if the same issue was already created by someone else.
+Before creating a new issue, be sure to try Troubleshooting and check if the same issue was already created by someone else.
 
 ## Resources ==(REQUIRED)==
 
-==*Do not change here! Keep as is or add only relevant documents/info for this example. Do not add any purchase link/marketing stuff*==
+==*Do not change here! Keep it as is or add only relevant documents/info for this example. Do not add any purchase link/marketing stuff*==
 
 * Official ESP32 Forum: [Link](https://esp32.com)
 * Arduino-ESP32 Official Repository: [espressif/arduino-esp32](https://github.com/espressif/arduino-esp32)
```
Comment (P-R-O-C-H-Y) [inline libraries/ESP32/examples/Template/ExampleTemplate/README.md:118]: @pedrominatel As this part was not edited and hidden in the PR, we should add ESP32-S3 datasheet link too. To have it in all future example READMEs.

**Pull request merged**